### PR TITLE
[com_fields] Migrate user groups from categories to field groups

### PR DESCRIPTION
**Pull request by laoneo**
Regression from PR #13019.

### Summary of Changes
This PR adapts the show custom fields setting in contacts to the new field groups.

### Testing Instructions
- Go to Users -> Field Groups
- Create two Groups with the name Demo1 and Demo2
- Go to Users -> Fields
- Create a field with the Title Demo Field 1 and assign it to the field group Demo1
- Create a field with the Title Demo Field 2 and assign it to the field group Demo2
- Create a field with the Title Demo Field and assign it to no field group
- Edit the super admin user and set some custom fields values
- Go to Components -> Contacts
- Open the options of the contacts component and select Demo1 in the parameter "Show User Custom Fields"
- Create a new contact and link it to the super admin user
- Create a new menu item for a single contact
- On the front open the new contact menu item

### Expected result
Only the values from the fields "Demo Field 1" and "Demo Field" should be shown.

### Actual result
The groups are not shown on the back end Contact options and on the front is the following warning displayed:

`Notice: Undefined property: stdClass::$catid in /joomla-cms/components/com_contact/views/contact/tmpl/default_user_custom_fields.php on line 24`

### CODE REVIEW

**Review by Bakual — DISMISSED** [review]
This PR looks good from review

**Comment by laoneo** [comment]
Yes I'v tested it up and down :wink:

**Comment by Bakual** [comment]
Is it expected that fields which belong to no group are shown as well when I select only "Group2"? Seems a bit counter-intuitive to me.

File: components/com_contact/views/contact/tmpl/default_user_custom_fields.php [hunk 1]
```diff
@@ -20,14 +20,14 @@
 	<?php return; ?>
 <?php endif; ?>
 
-<?php foreach ($this->contactUser->fields as $field) :?>
-	<?php if (!in_array('-1', $displayGroups) && !in_array($field->catid, $displayGroups)) : ?>
+<?php foreach ($this->contactUser->fields as $field) : ?>
+	<?php if (!in_array('-1', $displayGroups) && $field->group_id && !in_array($field->group_id, $displayGroups)) : ?>
```
Comment (Bakual) [inline components/com_contact/views/contact/tmpl/default_user_custom_fields.php]: Shouldn't that be
`<?php if (!in_array('-1', $displayGroups) && (!$field->group_id || !in_array($field->group_id, $displayGroups))) : ?>`

Skip if either group_id is 0 or it isn't in the displayGroup.

**Comment by laoneo** [comment]
Guess it is time to define what happens with fields which do not belong to a group. IMO they should be displayed. No group means for me that it belongs to all groups.

File: components/com_contact/views/contact/tmpl/default_user_custom_fields.php (hunk 1 above)
Comment (laoneo) [inline components/com_contact/views/contact/tmpl/default_user_custom_fields.php]: If it doesn't belong to no group, then display it.

**Comment by Bakual** [comment]
The tooltip indicates something else.
Also when nothing is selected, no fields are shown (as expected), but when I add a single group, actually two groups are shown. That's a bit unexpected to me.

**Comment by laoneo** [comment]
Which tooltip are you talking about?

**Comment by Bakual** [comment]
The one from the com_contact option
tooltip

**Comment by brianteeman** [comment]
There is the default group called fields if you dont put a field in a named group

**Comment by laoneo** [comment]
So if a field belongs to no group, should it be displayed or not? Current implementation is, it will be displayed as for me a field which doesn't belong to a group, it is some kind of in all groups (it starts to become strange as I'm typing this phrase). But I'm fine with whatever makes more sense.

**Comment by Bakual** [comment]
What I expect is if I select only one group, only the fields of that group should show.
If I select "All", it should also show the ones without any group.

File: components/com_contact/views/contact/tmpl/default_user_custom_fields.php [hunk 2]
```diff
@@ -20,14 +20,14 @@
...
-	<?php if (!key_exists($field->category_title, $userFieldGroups)) : ?>
-		<?php $userFieldGroups[$field->category_title] = array();?>
+	<?php if (!key_exists($field->group_title, $userFieldGroups)) : ?>
+		<?php $userFieldGroups[$field->group_title] = array();?>
 	<?php endif; ?>
-	<?php $userFieldGroups[$field->category_title][] = $field;?>
+	<?php $userFieldGroups[$field->group_title][] = $field;?>
 <?php endforeach; ?>
 
 <?php foreach ($userFieldGroups as $categoryTitle => $fields) :?>
```
Comment (Bakual) [inline components/com_contact/views/contact/tmpl/default_user_custom_fields.php]: Can you change `$categoryTitle` to `$groupTitle` here as well? while technically not needed it makes the code more understandable :smile: 
Same for the Beez override obviously.

**Comment by laoneo** [comment]
Done

**Comment by Bakual** [comment]
Thanks. :+1:

**Comment by brianteeman** [comment]
> Imho only fields from the selected group should be shown.

MY view is 
if you select ALL then all fields should display
if you select GROUP1 then only fields in that group should display

The problem comes with fields that are not in a group (these are my generic fields). 

So for example I have GROUP1, GROUP2 and some fields not in a group. How would I display the fields that are not in a group and GROUP1 but not GROUP2

**Comment by laoneo** [comment]
We can add another option to select like "No grouped fields", maybe, not sure, hmmm.

**Comment by brianteeman** [comment]
Or maybe we require fields to be in a groups just as we require articles etc to be in categories

**Comment by Bakual** [comment]
We could require the group. Or we could just say if you want to show those non-grouped fields here, create a group and group them 
Adding an option for not grouped fields imho only asks for troubles. It complicates the code and it's also not that simple to find a good option name.

**Comment by laoneo** [comment]
> Or we could just say if you want to show those non-grouped fields here, create a group and group them

Sometimes the answer is so easy. I would go that way.

**Comment by laoneo** [comment]
Changed code to show only fields of the selected groups.

**Comment by ralain** [comment]
I have tested this item :white_check_mark: successfully on 5b3ee16594d83ffd254ed05216915c1ef7fac3df<br /><br />Notice is gone, and specific field groups can be selected in com_contact options in addition to "All".<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13103">issues.joomla.org/tracker/joomla-cms/13103</a>.</sub>

**Comment by jeckodevelopment** [comment]
@laoneo could you please look at the conflicts?

**Comment by anibalsanchez** [comment]
Yes, I tested successfully most of the cases. Users and Contacts with several Field Groups and Fields (including fields with no group).

.... But "Show User Custom Fields" only shows All... What am I missing?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13103">issues.joomla.org/tracker/joomla-cms/13103</a>.</sub>

**Comment by laoneo** [comment]
You can control which user custom fields should be shown on the contact view on the front end based on the field groups.

**Comment by anibalsanchez** [comment]
Yes, I know... but it only allows to select "All"

**Comment by anibalsanchez** [comment]
test-all

**Comment by laoneo** [comment]
Did you create some User (not contact) field groups?

**Comment by anibalsanchez** [comment]
Yep, I created Field Groups and Fields in Users and Contacts. They work perfectly well.

Which is the method that populates this field?

**Comment by laoneo** [comment]
It's this one https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_fields/models/fields/fieldgroups.php

**Comment by laoneo** [comment]
@jeckodevelopment done. Can you please test it, would be great to have it in the first alpha?

**Comment by laoneo** [comment]
@anibalsanchez I guess I found the issue. Because of PR #13175 the identifier changed from extension to context. I'v fixed it in my last commit. Now it should work again as expected.

**Comment by anibalsanchez** [comment]
@laoneo The query in JFormFieldFieldgroups requires an extension parameter. But, it is not defined in config.xml.

I've just added this line and it works OK:

	extension="com_users"

**Comment by laoneo** [comment]
It depends in the status of your branch. If yoi have the most actual code with #13175 which was merged recently then you need the context variable. If it is older, them you need the extension one. Anyway please mark it as successfull test. Thanks.

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 2b71bd8a1f0b6a9a0db6bf907ab6b9a4cea6b499<br /><br />Test OK!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13103">issues.joomla.org/tracker/joomla-cms/13103</a>.</sub>

**Comment by ralain** [comment]
I have tested this item :white_check_mark: successfully on 2b71bd8a1f0b6a9a0db6bf907ab6b9a4cea6b499<br /><br />Test still good.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13103">issues.joomla.org/tracker/joomla-cms/13103</a>.</sub>

**Comment by jeckodevelopment** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13103">issues.joomla.org/tracker/joomla-cms/13103</a>.</sub>

**Pull request merged**